metadatastore/commands.py

```diff
@@ -39,167 +99,512 @@ def inner(*args, **kwargs):
+        a Document/dict with a 'uid' key or a uid string
+
+    Returns
+    -------
+    runstop : doc.Document
+        The RunStop document
+
+    Raises
+    ------
+    NoRunStop
+        If no RunStop document exists for the given RunStart
+
+    """
+
+    runstart_uid = doc_or_uid_to_uid(runstart)
+    oid = _RUNSTART_UID_to_OID_MAP[runstart_uid]
+
+    runstop = RunStop._get_collection().find_one(
+        {'run_start_id': oid})
+
+    if runstop is None:
+        raise NoRunStop("No run stop exists")
+
+    return _cache_runstop(runstop)
+
+
+def descriptors_by_runstart(runstart):
+    """Given a RunStart return a list of it's descriptors
+
+    Raises if no EventDescriptors exist.
+
+    Parameters
+    ----------
+    runstart : doc.Document or dict or str
+        The RunStart to get the RunStop for.  Can be either
+        a Document/dict with a 'uid' key or a uid string
+
+    Returns
+    -------
+    event_descriptors : list
+        A list of EventDescriptor documents
+
+    Raises
+    ------
+    NoEventDescriptors
+        If no EventDescriptor documents exist for the given RunStart
+    """
+    # normalize the input and get the runstart oid
+    runstart_uid = doc_or_uid_to_uid(runstart)
+    oid = _RUNSTART_UID_to_OID_MAP[runstart_uid]
+
+    # query the database for any event descriptors which
+    # refer to the given runstart
+    descriptors = EventDescriptor._get_collection().find(
+        {'run_start_id': oid})
+
+    # loop over the found documents, cache, and dereference
+    rets = [_cache_descriptor(descriptor) for descriptor in descriptors]
+
+    # if nothing found, raise
+    if not rets:
+        raise NoEventDescriptors("No EventDescriptors exists")
+
+    # return the list of event descriptors
+    return rets
+
+
+def fetch_events_generator(descriptor):
+    """A generator which yields all events from the event stream
+
+    Parameters
+    ----------
+    descriptor : doc.Document or dict or str
+        The RunStart to get the RunStop for.  Can be either
+        a Document/dict with a 'uid' key or a uid string
+
+    Yields
+    ------
+    event : doc.Document
+        All events for the given EventDescriptor from oldest to
+        newest
+    """
+    descriptor_uid = doc_or_uid_to_uid(descriptor)
+    descriptor = descriptor_given_uid(descriptor_uid)
+    col = Event._get_collection()
+
+    oid = _EVENTDESC_UID_to_OID_MAP[descriptor_uid]
+
+    ev_cur = col.find({'descriptor_id': oid},
+                      sort=[('time', 1)])
+
+    for ev in ev_cur:
+        # ditch the ObjectID
+        ev.pop('_id')
+        # pop the descriptor oid
+        ev.pop('descriptor_id')
+        # replace it with the defererenced descriptor
+        ev['descriptor'] = descriptor
+        # pop the data
+        data = ev.pop('data')
+        # replace it with the friendly paired dicts
+        ev['data'], ev['timestamps'] = [{k: v[j] for k, v in data.items()}
```

Comment:
Did some benchmarking with retrieval of 10k events and the two are no different in speed. I concur that you should leave it as `data = ev.pop('data')` and then iterate over data.items